### PR TITLE
Add a KDF password manager.

### LINKED ISSUE #9: Implement a pbkdf password manager?
This is a feature of bcrypt 3.0.

### PR DESCRIPTION
**Pull request by jamadden**
Fixes #9

Also refactor some of the internals of password.py to remove accumulated duplication in match() methods and encoding.

### CODE REVIEW

File: src/zope/password/password.py [hunk 1]
```diff
@@ -492,14 +491,13 @@ def checkPassword(self, encoded_password, password):
         return encoded_password == self.encodePassword(password)[5:]
 
     def match(self, encoded_password):
-        if not isinstance(encoded_password, bytes):
-            encoded_password = encoded_password.encode('ascii')
+        encoded_password = _encoder(encoded_password)
         return (
-            encoded_password.startswith(b'{SHA}') or
+            encoded_password.startswith(self._prefix) or
             encoded_password.startswith(b'{SHA1}'))
```
Comment (mgedmin) [inline src/zope/password/password.py]: You could shorten this to a single `.startswith((self._prefix, b'{SHA1}'))` call.

File: src/zope/password/password.py [hunk 2]
```diff
@@ -582,9 +571,105 @@ def match(self, hashed_password):
...
         """
+        hashed_password = _encoder(hashed_password)
         return (hashed_password.startswith(self._prefix)
                 or self._z3c_bcrypt_syntax.match(hashed_password) is not None)
 
+class BCRYPTKDFPasswordManager(_PrefixedPasswordManager):
+    """
+    BCRYPT KDF password manager.
+
+    This manager converts a plain text password into a byte array .
```
Comment (mgedmin) [inline src/zope/password/password.py]: Strange space in front of the `.`.

File: src/zope/password/password.py [hunk 3]
```diff
@@ -582,9 +571,105 @@ def match(self, hashed_password):
...
+    And the old password still checks out:
+
+    >>> manager.checkPassword(encoded, password)
+    True
+    """
+
+    #: The number of rounds of hashing that should be applied.
+    #: The higher the number, the slower it is. It should be at least
+    #: 50.
+    rounds = 1<<10
```
Comment (mgedmin) [inline src/zope/password/password.py]: I'm wondering if `rounds = 2**10` wouldn't be a bit more Pythonic.

File: src/zope/password/password.py [hunk 4]
```diff
@@ -582,9 +571,105 @@ def match(self, hashed_password):
...
+        rounds, salt, key = hashed_password[len(self._prefix):].split(b'$')
+        rounds = int(rounds, 16)
+
+        salt = urlsafe_b64decode(salt)
+        keylen = len(urlsafe_b64decode(key))
+        encoded_password = self._encode(clear_password, salt, rounds, keylen)
+        # XXX: Do we need to use a timing-safe comparison function here?
+        # z3c.bcrypt does, and so does bcrypt.hashpw. But there is no such
+        # public function
+        return hashed_password == encoded_password
```
Comment (mgedmin) [inline src/zope/password/password.py]: [hmac.compare_digest()](https://docs.python.org/3/library/hmac.html#hmac.compare_digest) is such a function, but only available in Python starting with 3.3.

File: src/zope/password/password.py [hunk 5]
```diff
@@ -582,9 +571,105 @@ def match(self, hashed_password):
...
+    def encodePassword(self, password):
+        salt = bcrypt.gensalt()
+        return self._encode(password, salt, self.rounds, self.keylen)
+
+    def checkPassword(self, hashed_password, clear_password):
+        hashed_password = _encoder(hashed_password)
+        if not self.match(hashed_password):
+            return False
+        rounds, salt, key = hashed_password[len(self._prefix):].split(b'$')
+        rounds = int(rounds, 16)
```
Comment (mgedmin) [inline src/zope/password/password.py]: I'm worrying that it's easy to make this function crash if you provide corrupt hashes (e.g. wrong number of '$' signs or non-hex digits in the `rounds` portion).

I am not a crypto person and I don't know whether that's fine (hashes are probably not attacker-supplied?) or not.

File: src/zope/password/password.py (hunk 1 above)
Comment (jamadden) [reply src/zope/password/password.py]: Thanks. I always forget that. Possibly because if you leave off the extra set of parentheses, bad things happen.

File: src/zope/password/password.py (hunk 2 above)
Comment (jamadden) [reply src/zope/password/password.py]: Whoops. Thanks.

File: src/zope/password/password.py (hunk 3 above)
Comment (jamadden) [reply src/zope/password/password.py]: Or even just the integer 1024. 

I just did what z3c.bcrypt did, although I reduced the rounds a bit. The original rounds was taking something like four seconds on my newish-computer, which seems excessive.

File: src/zope/password/password.py (hunk 5 above)
Comment (jamadden) [reply src/zope/password/password.py]: We want to avoid false positives here, so a crash is better than that, I think. But the hashed password would not be supplied by the attacher, the `clear_password` is. The hashed password is stored by the system and previously generated by this object, if it's being used according to the contract of `IMatchingPasswordManager`.

File: src/zope/password/password.py (hunk 4 above)
Comment (jamadden) [reply src/zope/password/password.py]: Thanks, I'll add that to the comment.

Comment (jamadden) [reply src/zope/password/password.py]: Actually, I'll just use it when available. Thanks.

Comment (jamadden) [reply src/zope/password/password.py]: Interestingly, `hmac.compare_digest` was actually [backported to 2.7.7](https://github.com/python/cpython/commit/629026aecc6bbb7cf8df25f7699eaf97d8e45701)  in 2014.  Do we need to care about supporting older point releases in a new version? I can add appropriate `requires-python` metadata to make sure a new release is not installed by older versions.

Comment (jamadden) [reply src/zope/password/password.py]: > Do we need to care about supporting older point releases in a new version?

Given that this is a security feature for a security-critical package, and that people who care about security are probably running 2.7.9+ anyway, I feel like it is going to be OK to require 2.7.7+ (a three-year-old release) in a new version. Especially since it adds new security-related features. And especially especially since the requires-python metadata generally means it won't get installed by accident. 

I added this to the PR.

**Comment by jamadden** [comment]
Thanks!
